File: TicTacToeApp/Server/AP1_Jv_T03.ID_1369281-1/src/Project_03/src/main/java/org/datasource/repository/GameRepository.java
package org.datasource.repository;

import org.datasource.model.GameEntity;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface GameRepository {

    /**
     * Сохранение текущей игры
     * @param gameEntity игра для сохранения
     * @return сохраненная игра
     */
    GameEntity save(GameEntity gameEntity);

    /**
     * Получение текущей игры по ID
     * @param id идентификатор игры
     * @return Optional с игрой или пустой, если игра не найдена
     */
    Optional<GameEntity> findById(UUID id);

    /**
 * Найти все игры.
 * @return Список всех GameEntity.
 */
List<GameEntity> findAll(); 
}

File: TicTacToeApp/app/src/main/java/com/example/tictactoeapp/domain/repository/GameRepository.kt
package com.example.tictactoeapp.domain.repository

import com.example.tictactoeapp.domain.model.Game
import com.example.tictactoeapp.domain.model.GameHistory
import com.example.tictactoeapp.domain.model.Leaderboard
import io.reactivex.Single
import io.reactivex.Completable
import io.reactivex.Maybe
import java.util.*

interface GameRepository {
    fun startNewGame(): Single<Game>
    fun makeMove(gameId: UUID, game: Game): Single<Game>
    fun getGame(gameId: UUID): Single<Game>
    fun saveGameLocally(game: Game): Completable
    fun getLocalGame(gameId: UUID): Maybe<Game>
    fun getAvailableGames(): Single<List<Game>>
    fun joinGame(gameId: UUID): Single<Game>
    fun startNewGameWithComputer(): Single<Game>
    fun startNewGameWithPlayer(): Single<Game>
    fun playerLeftGame(gameId: UUID): Single<Game>
    fun getGameHistory(): Single<List<GameHistory>>
    fun getLeaderboard(): Single<List<Leaderboard>>
}